### PR TITLE
Implement compression

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

File: ws/ws.nim [hunk 1]
```diff
@@ -260,7 +304,7 @@ proc handleRequest*(
 proc new*(
   _: typedesc[WSServer],
   protos: openArray[string] = [""],
-  factories: openArray[ExtFactory] = [],
+  factories: openArray[(string, ExtFactory)] = [],
```
Comment (dryajov) [inline ws/ws.nim:307]: Was thinking about this. Maybe just pass a `Table` instead (not sure)?

**Comment by dryajov** [comment]
`ExtParam` should probably be moved to `types.nim`.

File: ws/ws.nim [hunk 2]
```diff
@@ -183,6 +183,45 @@ proc connect*(
     onClose = onClose,
     rng = rng)
 
+proc selectExt(ws: WSServer, request: HttpRequest): string
+  {.raises: [Defect, WSExtError].} =
+
+  # it is possible to have multiple "Sec-WebSocket-Extensions"
+  let exts = request.headers.getList("Sec-WebSocket-Extensions")
+  var extList: seq[AppExt]
```
Comment (dryajov) [inline ws/ws.nim:191]: I think `AppExt` is the same as `Ext`, should we merge them?

File: ws/types.nim [hunk 3]
```diff
@@ -92,10 +92,9 @@ type
 
   Ext* = ref object of RootObj
     name*: string
-    options*: Table[string, string]
 
-  ExtFactory* = proc(name: string, options: Table[string, string]):
-    Ext {.raises: [Defect].}
+  ExtFactory* = proc(serverSide: bool, options: seq[ExtParam]):
```
Comment (dryajov) [inline ws/types.nim:96]: Do we really need to make a distinction for server vs client side?

File: ws/ws.nim (hunk 1 above)
Comment (jangko) [reply ws/ws.nim:307]: I have strong reason to keep the `openArray` interface.

1. not sacrifice ergonomics. I imagine authors of extensions would export their extensions using:
```Nim
const
    deflateExt* = ("permessage-deflate", deflateFactory)

const
    snappyExt* = ("permessage-snappy", snappyFactory)

etc..

let ws = WSServer.new(factories = [deflateExt, snappyExt])  # still ergonomic
```

2. flexibility to swap internal implementation
I assume in real life production code, the number of extensions is not very much. probably not much than 5.
If that is the case, it is more economically to use plain `seq` instead of `Table`. Table is overkill if number of exts is small.
keeping the `openArray` interface will allow us to change the implementation without breaking the interface.

File: ws/ws.nim (hunk 2 above)
Comment (jangko) [reply ws/ws.nim:191]: I agree ExtParam should be moved to `types.nim`.

My idea to separate `AppExt` and `Ext` is to separate server code(ext creation) and parser code. If we want to merge them, then the parser code and server code will be interleaving.  But I will revisit this later to find a clean solution.

File: ws/types.nim (hunk 3 above)
Comment (jangko) [reply ws/types.nim:96]: currently I'm using it for optimization and diagnosis purpose, I need more thorough analysis before discard it.

File: ws/ws.nim (hunk 1 above)
Comment (dryajov) [reply ws/ws.nim:307]: > I have strong reason to keep the openArray interface.

OK, I like the reasoning and I don't really have a strong objection against this.

> If that is the case, it is more economically to use plain seq instead of Table. Table is overkill if number of exts is small.

Yeah agree with that, I believe I had it as a `seq` initially?

File: ws/ws.nim (hunk 2 above)
Comment (dryajov) [reply ws/ws.nim:191]: I think that for the most part, the client/server functionality is very similar, if we can merge them I'd prefer we do that.

File: ws/ws.nim (hunk 1 above)
Comment (jangko) [reply ws/ws.nim:307]: yes, and I will change it back to seq.

**Comment by jangko** [comment]
@dryajov , the server side compression is almost  ready. maybe you can review it again while I prepare the miniz repo.
for the client side compression, that will be in the next PR.

File: ws/ws.nim [hunk 4]
```diff
@@ -183,6 +183,64 @@ proc connect*(
     onClose = onClose,
     rng = rng)
 
+proc extInList(ws: WSServer, extName: string): bool =
```
Comment (dryajov) [inline ws/ws.nim:186]: `contains` would be more idiomatic here

File: ws/compression.nim [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  for i in 0..<arg.value.len:
+    winbit = winbit * 10 + arg.value[i].int - '0'.int
+
+  if winbit < 8 or winbit > 15:
+    return err("window bits should between 8-15, got " & $winbit)
+
+  res = winbit
+  return ok("=" & arg.value)
+
+proc validateParams(args: seq[ExtParam],
```
Comment (dryajov) [inline ws/compression.nim:60]: Would something like `createParams` reflect it's intent better?

**Comment by dryajov** [comment]
> @dryajov , the server side compression is almost ready. maybe you can review it again while I prepare the miniz repo.
> for the client side compression, that will be in the next PR.

@jangko I spent some time looking into how extensions and specifically per message deflate works and this PR needs to be reworked to some extent. 

The core of the issue is that extensions can/should hijack the flow, as it is the case of the per message deflate as outlined here - https://datatracker.ietf.org/doc/html/rfc7692#section-6:

- Compressing implies that the message is compressed and then passed to the next extension in the chain, per message deflate is already the last extension, then it will frame the compressed message.
- Decompressing implies buffering the message, decompressing it and then passing the body and/or frame to the next extension in the chain or, the application, if per message deflate is the last extension.

The way it is implemented now, won't work 100% of the cases. For example calling `decode(...)` [here](https://github.com/status-im/nim-ws/pull/49/files#diff-06f7698d7319a084ba0ec1981ccdaa2f650026e437b179ab5afe22e5f67745b9R352) is the wrong thing to do as this is only a convenience proc to avoid having to deal with pointers and such.

In short we need to rework some of the flow to accommodate for this requirements. I don't think this would affect the core functionality of this PR and we can rework the websockets flow in isolation. Once that's done, we can rebase and merge this changes.

**Review by dryajov — CHANGES_REQUESTED** [review]
I'll leave it blocked for now to get the required changes to the extensions flow before merging this one in.

**Comment by jangko** [comment]
actually my current approach is not working at all. 
the spec not only require specific flow in websocket implementation, but also specific way *to use* the compression library which is quite confusing initially -- and take significant amount of time to understand what is going on.

1. after some experiments, it seems your original approach to the extension is already correct. -- compression done at per frame basis, compression flushing is important.
2. the websocket compression is designed around zlib interface and behavior. we can only use compression lib which support zlib behavior. in this case we still can use miniz. 
3. something like https://github.com/status-im/nim-zippy is unusable without serious modification.
    - it doesn't support LZ77 window size adjustment
    - it doesn't support segmented buffer for compression and decompression. all must done in one go.

**Comment by dryajov** [comment]
> after some experiments, it seems your original approach to the extension is already correct. -- compression done at per frame basis, compression flushing is important. and compressor lib must support segmented buffer.

Yes, this is what I noticed as well, fortunately at least at the websockets level it doesn't seem to hard to modify `WSSession` and `Frame` to do the right thing.

>  we can only use compression lib which support zlib behavior. in this case we still can use miniz.

OK, good to know we can still use miniz - is it already wrapped? If not would zlib be more appropriate?

https://github.com/status-im/nim-zippy seems to be a pure Nim implementation - how hard would it be to extend it to support lz77 windowing?

**Comment by jangko** [comment]
we already have https://github.com/status-im/nim-miniz.

> how hard would it be to extend it to support lz77 windowing?

easy.

but the segmented buffer and flushing mechanism is hard, because we need to redesign many of it's internal flow from contiguous operation into `pause-resume` operation.

File: ws/ws.nim (hunk 4 above)
Comment (jangko) [reply ws/ws.nim]: fixed

File: ws/compression.nim (hunk 5 above)
Comment (jangko) [reply ws/compression.nim]: fixed

**Comment by jangko** [comment]
the server side compression now works.
it pass all autobahn tests section 12 and 13 cleanly.
no test cases because i still need to implement client compression in next PR.
we still cannot test client using autobahn due to #32.

**Comment by dryajov** [comment]
> the server side compression now works.
> it pass all autobahn tests section 12 and 13 cleanly.
> no test cases because i still need to implement client compression in next PR.
> we still cannot test client using autobahn due to #32.

I believe `Host` is optional, why is it blocking the tests?

**Comment by jangko** [comment]
> I believe Host is optional, why is it blocking the tests?

Fixes in this branch https://github.com/status-im/nim-ws/tree/auto_client allow autobahn_client to run. It sits on top of your PR(#56).

Don't know why autobahn server need 'Host' field. perhaps a bug.

File: ws/ws.nim [hunk 6]
```diff
@@ -250,17 +313,23 @@ proc handleRequest*(
     raise newException(WSHandshakeError,
         "Failed to sent handshake response. Error: " & exc.msg)
 
-  return WSSession(
+  let session = WSSession(
     readyState: ReadyState.Open,
     stream: request.stream,
     proto: protocol,
+    extensions: system.move(ws.extensions),
```
Comment (dryajov) [inline ws/ws.nim:320]: I think `extensions` is shared across all sessions in server mode?

Comment (dryajov) [reply ws/ws.nim:320]: I guess I'm a bit confused about the `move` here, why do we need it?

Comment (jangko) [reply ws/ws.nim:320]: no, it is not shared. the server will create new one for each session.

Comment (jangko) [reply ws/ws.nim:320]: for extension such as `permessage-deflate`, as it's name implies. will keep a de/compression context from the start of message to the end of message. from first frame to last frame will use the same context. so extensions cannot be shared among session.

Comment (jangko) [reply ws/ws.nim:320]: for a simple extension without configuration parameter, may the factory can keep a singleton to be shared among session.
but again, the extension chain itself can be different from one session to the next.

Comment (dryajov) [reply ws/ws.nim:340]: > for extension such as permessage-deflate, as it's name implies. will keep a de/compression context from the start of message to the end of message. from first frame to last frame will use the same context. so extensions cannot be shared among session.

OK right, that makes sense then.

**Comment by jangko** [comment]
current status of this PR:
- server side compression already pass all autobahn tests
- client side compression have many failing test cases because of:
   * autobahn server reject our extensions offer and force us to use unsupported compression feature(not supported by miniz).
   * succesfully negotiated compression configuration failing for mysterious reason. what I don't understand is the same test data    is used by both server and client mode, but the result/behavior is different.
  
will experiment more using zlib and see what is actually going on.

**Comment by dryajov** [comment]
> not supported by miniz

Would it make sense to reconsider miniz in favor of a more complete implementation, or even just a native nim one? I much rather  have control over the entirety of the compression stack specially when already have a partial nim implementation.

**Comment by jangko** [comment]
sure, we need to find a more complete deflate lib. but so far only zlib the strongest candidate. other nim deflate libs are not suitable or need too much modification.

**Comment by jangko** [comment]
finally we have a working compression extension for both client and server.
